### PR TITLE
ffi: Extract IR unit deserialization methods from `Deserializer` into standalone functions to prepare for a templatized `Deserializer`.

### PR DESCRIPTION
**Pull request by LinZhihao-723**
# Description

This PR is working towards #539 

The workflow of the new deserializer API as discussed in #539 will have the following steps:
1. Check the IR unit type
2. Call the corresponding IR unit deserialization method
3. Apply the result to update the underlying state if necessary
4. Pass the result from 2 to user-defined ir unit handler

Before this PR, step 2 & 3 are in the same method. But thinking the deserializer as a high-level IR unit processor, it's cleaner to split these methods out into different methods. In an ideal implementation, we should make these methods as private methods of the deserializer. However, the deserializer will be implemented as a template to receive an IR unit handler defined in #540, which means we need to implement all the methods from step 2 to step 3 in the header file if we want to make them as private members of the deserializer. This will make the header too messy to manage.

As a solution for now, we decided to split the methods from step 2 and move them into a new set of the files named "ir_unit_deserialization_methods". This allows us to: (1). make things cleaner according to the workflow discussed above; (2). split most of the low-level deserialization into an implementation file. The downside of this solution is that these low-level APIs will be exposed directly to users as they are public C-style APIs. In the future, we should formally compile FFI core as a library and decide what public headers users can include.

For now, we use these new methods to re-implement the current `deserialize_to_next_log_event` method to consume the IR stream unit by unit.

# Validation performed

- Ensure the related unit tests still passed
- Ensure all workflow passed

## Summary by CodeRabbit

- **New Features**
	- Introduced new functions for deserializing Intermediate Representation (IR) units, enhancing schema tree and log event handling.
	- Added error handling capabilities to convert internal error codes to standard error codes.

- **Bug Fixes**
	- Updated deserialization logic to handle End-of-File (EOF) conditions, ensuring appropriate error responses.

- **Documentation**
	- Improved comments in the `Deserializer` class for better clarity on method functionalities. 

- **Tests**
	- Enhanced test cases to verify deserialization behaviour under EOF conditions.

### CODE REVIEW

File: components/core/src/clp/ffi/ir_stream/ir_unit_deserialization_methods.cpp [hunk 1]
```diff
@@ -0,0 +1,535 @@
...
+        case cProtocol::Payload::SchemaTreeNodeStr:
+            return SchemaTreeNode::Type::Str;
+        case cProtocol::Payload::SchemaTreeNodeUnstructuredArray:
+            return SchemaTreeNode::Type::UnstructuredArray;
+        case cProtocol::Payload::SchemaTreeNodeObj:
+            return SchemaTreeNode::Type::Obj;
+        default:
+            return std::nullopt;
+    }
+}
```
[the remark this answers is not in the document]
Comment (LinZhihao-723) [reply components/core/src/clp/ffi/ir_stream/ir_unit_deserialization_methods.cpp:186]: Only trivially destructible global objects are allowed in our coding standard.

File: components/core/src/clp/ffi/ir_stream/ir_unit_deserialization_methods.cpp [hunk 2]
```diff
@@ -0,0 +1,535 @@
...
+        int64_t deserialized_val{};
+        if (false == deserialize_int(reader, deserialized_val)) {
+            return IRErrorCode::IRErrorCode_Incomplete_IR;
+        }
+        val = deserialized_val;
+    } else {
+        return IRErrorCode::IRErrorCode_Corrupted_IR;
+    }
+    return IRErrorCode::IRErrorCode_Success;
+}
```
[the remark this answers is not in the document]
Comment (LinZhihao-723) [reply components/core/src/clp/ffi/ir_stream/ir_unit_deserialization_methods.cpp:261]: The methods won't simplify the logic inside `deserialize_int_val` as we still need to determine the integer type during the runtime.

File: components/core/src/clp/ffi/ir_stream/ir_unit_deserialization_methods.cpp [hunk 3]
```diff
@@ -0,0 +1,535 @@
...
+            node_id_value_pairs.emplace(node_id, Value{});
+            break;
+        case cProtocol::Payload::ValueEmpty:
+            node_id_value_pairs.emplace(node_id, std::nullopt);
+            break;
+        default:
+            return IRErrorCode::IRErrorCode_Corrupted_IR;
+    }
+    return IRErrorCode::IRErrorCode_Success;
+}
```
[the remark this answers is not in the document]
Comment (LinZhihao-723) [reply components/core/src/clp/ffi/ir_stream/ir_unit_deserialization_methods.cpp:393]: We've added the check on the caller end already.

**Review by davidlion — CHANGES_REQUESTED** [review]
Code looks good to me. Lets just fix the doc strings and update the PR description regarding the long-term plan for public APIs as we discussed.

**Review by davidlion — APPROVED** [review]

**Pull request merged**